rs/cli/src/cordoned_feature_fetcher.rs

```diff
@@ -0,0 +1,174 @@
+use std::{path::PathBuf, time::Duration};
+
+use decentralization::network::NodeFeaturePair;
+use futures::future::BoxFuture;
+use ic_management_types::NodeFeature;
+use itertools::Itertools;
+use log::warn;
+use mockall::automock;
+use reqwest::{Client, ClientBuilder};
+use strum::VariantNames;
+
+#[automock]
+pub trait CordonedFeatureFetcher: Sync + Send {
+    fn fetch(&self) -> BoxFuture<'_, anyhow::Result<Vec<NodeFeaturePair>>>;
+}
+
+pub struct CordonedFeatureFetcherImpl {
+    client: Client,
+    fallback_file: Option<PathBuf>,
+    offline: bool,
+}
+
+const CORDONED_FEATURES_FILE_URL: &str = "https://raw.githubusercontent.com/dfinity/dre/refs/heads/main/cordoned_features.yaml";
+
+impl CordonedFeatureFetcherImpl {
+    pub fn new(offline: bool, fallback_file: Option<PathBuf>) -> anyhow::Result<Self> {
+        let client = ClientBuilder::new().timeout(Duration::from_secs(10)).build()?;
+
+        Ok(Self {
+            client,
+            fallback_file,
+            offline,
+        })
+    }
+
+    async fn fetch_from_git(&self) -> anyhow::Result<Vec<NodeFeaturePair>> {
+        let bytes = self
+            .client
+            .get(CORDONED_FEATURES_FILE_URL)
+            .send()
+            .await?
+            .error_for_status()?
+            .bytes()
+            .await?;
+
+        self.parse(&bytes)
+    }
+
+    fn fetch_from_file(&self, path: &PathBuf) -> anyhow::Result<Vec<NodeFeaturePair>> {
+        let contents = std::fs::read(path)?;
+
+        self.parse(&contents)
+    }
+
+    // Write tests for this
+    fn parse(&self, contents: &[u8]) -> anyhow::Result<Vec<NodeFeaturePair>> {
```

Comment:
This function doesn't seem to need `&self`.

Reply:
True! What I like about all these cases is that I can do `self.parse(...)` rather than `Self::parse()`. It makes the code more readable to me so that is why I prefer it. If you would rather me remove it all do it 😄

Reply:
It's one less argument that needs to be passed to the function, but it's not a big deal either way. We are not performance sensitive in this call. So it's fine if we leave it as is.